### PR TITLE
replace record "with clause" with record spread operator

### PR DESCRIPTION
**Pull request by mccanne**

### CODE REVIEW

File: compiler/ast/ast.go [hunk 1]
```diff
@@ -91,16 +91,29 @@ type RegexpSearch struct {
...
+}
+
+type RecordElem interface {
+	recordAST()
 }
 
-type FieldExpr struct {
+func (*ID) recordAST()     {}
+func (*Field) recordAST()  {}
+func (*Spread) recordAST() {}
```
Comment (nwt) [inline compiler/ast/ast.go:104]:
```suggestion
func (*Field) recordAST()  {}
func (*ID) recordAST()     {}
func (*Spread) recordAST() {}
```

File: expr/values.go [hunk 2]
```diff
@@ -55,98 +65,117 @@ func (r *RecordExpr) Eval(ectx Context, this *zed.Value) *zed.Value {
...
+			rec := elem.Spread.Eval(ectx, this)
+			if rec.IsMissing() {
+				continue
+			}
+			typ := zed.TypeRecordOf(rec.Type)
+			if typ == nil {
+				// Treat non-record spread values like missing.
+				continue
+			}
+			it := rec.Bytes.Iter()
```
Comment (nwt) [inline expr/values.go:108]:
```suggestion
			it := rec.Iter()
```

File: expr/ztests/record-spread-var.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+zed: |
+  over a with b => ( yield {a:this+10,b} )
```
Comment (nwt) [inline expr/ztests/record-spread-var.yaml:2]: Did you mean to use the spread operator here?

File: expr/values.go [hunk 4]
```diff
@@ -55,98 +65,117 @@ func (r *RecordExpr) Eval(ectx Context, this *zed.Value) *zed.Value {
...
+		r.bytes[field.colno] = field.value.Bytes
 	}
-	if colno < len(cols) {
-		dirty = true
-		cols = cols[:colno]
+}
+
+func (r *recordSpreadExpr) invalidate(object map[string]column) {
+	n := len(object)
+	if cap(r.columns) < n {
```
Comment (nwt) [inline expr/values.go:167]:
```suggestion
	if n := len(object); cap(r.columns) < n {
```

File: expr/values.go [hunk 5]
```diff
@@ -55,98 +65,117 @@ func (r *RecordExpr) Eval(ectx Context, this *zed.Value) *zed.Value {
...
+// seeing if we can reuse the cached record type.  If not we look up
+// a new type, cache it, and save the columns for the cache check.
+func (r *recordSpreadExpr) update(object map[string]column) {
+	if len(r.columns) != len(object) {
+		r.invalidate(object)
+		return
+	}
+	for name, field := range object {
+		col := zed.Column{name, field.value.Type}
+		if r.columns[field.colno] != col {
```
Comment (nwt) [inline expr/values.go:157]:
```suggestion
		if r.columns[field.colno] != zed.Column{name, field.value.Type} {
```

File: compiler/kernel/expr.go [hunk 6]
```diff
@@ -364,24 +364,27 @@ func compileRegexpMatch(zctx *zed.Context, match *dag.RegexpMatch) (expr.Evaluat
...
+	var elems []expr.RecordElem
+	for _, elem := range record.Elems {
+		switch elem := elem.(type) {
+		case *dag.Spread:
+			e, err := compileExpr(zctx, elem.Expr)
+			if err != nil {
+				return nil, err
+			}
+			elems = append(elems, expr.RecordElem{Spread: e})
+		case *dag.Field:
```
Comment (nwt) [inline compiler/kernel/expr.go:376]: Nit: Maybe swap the case order?

**Review by nwt — APPROVED** [review]

File: expr/ztests/record-spread-var.yaml (hunk 3 above)
Comment (mccanne) [reply expr/ztests/record-spread-var.yaml:2]: I'll change the name...

File: expr/values.go (hunk 5 above)
Comment (mccanne) [reply expr/values.go:157]: I got a syntax error when I tried that.

**Pull request merged**